services/supervisor/src/simple/mod.rs

```diff
@@ -22,17 +22,90 @@ use exonum::{
+};
 
 mod api;
 mod schema;
 pub use self::schema::Schema;
 
+/// Applies configuration changes, isolating each of them with by using `Fork` checkpoints.
+///
+/// # Safety
+///
+/// This function should be used with extreme care. It makes the following assumptions:
+///
+/// - The function must be called at the end of the transaction or `before_commit` execution.
+///   If the transaction errors / panics afterwards, the changes to the configs will not
+///   be rolled back.
+/// - No changes to the blockchain state should be introduced before the call to this function.
+///   Any changes that are introduced will be committed regardless of the execution status,
+///   or the status of application of any config change. This is if the execution wasn't interrupted
+///   by a panic / error *before* hitting the call; if this happens, the usual rules apply.
+///
+/// These restrictions are the result of `Fork` not having multi-layered checkpoints.
+fn update_configs(context: &mut CallContext<'_>, changes: Vec<ConfigChange>) {
+    // An error while configuring one of the service instances should not affect others.
+    changes.into_iter().for_each(|change| match change {
+        ConfigChange::Consensus(config) => {
+            log::trace!("Updating consensus configuration {:?}", config);
+            let configure_result = apply_consensus_config(context, config);
+
+            if let Err(e) = configure_result {
+                // Consensus configuration is not allowed to fail, since it was validated and approved.
+                // This is an unrecoverable error.
+                panic!("Consensus configuration failed with error {:?}", e)
+            }
+        }
+
+        ConfigChange::Service(config) => {
+            log::trace!(
+                "Updating service instance configuration, instance ID is {}",
+                config.instance_id
+            );
+            let configure_result = apply_service_config(context, config);
+
+            if let Err(e) = configure_result {
+                log::error!(
+                    "An error occurred while applying service configuration. {}",
+                    e
+                );
+            }
+        }
+
+        ConfigChange::DeployRequest(deploy_request) => {
+            log::trace!("Processing deploy request {:?}", &deploy_request);
+            // Activation height is ignored for simple supervisor.
+            let start_deploy_result =
+                context.start_artifact_registration(deploy_request.artifact, deploy_request.spec);
+            if let Err(e) = start_deploy_result {
+                // Deployment failures are not allowed within simple supervisor.
+                panic!("Deployment failed with error {:?}", e);
+            }
+        }
+
+        ConfigChange::StartService(start_request) => {
+            log::trace!("Processing start service request {:?}", &start_request);
+            let start_instance_result = context.start_adding_service(
```

Comment:
It is simpleier to use `expect` here in my opinion.

Reply:
`expect` is less visible IMO (usually `expect` means "Hm, I'm sure that here everything should be OK").
Here I wanted this `panic` to be visible as much as possible: we **can** fail and it will result in explicit panic.